File: android-core/src/test/java/com/mparticle/identity/IdentityApiTest.java
package com.mparticle.identity;

import com.mparticle.internal.MessageManager;
import com.mparticle.mock.AbstractMParticleUser;
import com.mparticle.testutils.AndroidUtils;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class IdentityApiTest {
    Random ran = new Random();

    //IdentityApi should order their users from greatest lastSeenTime to least
    @Test
    public void testUserOrder() {
        IdentityApi identityApi = new IdentityApi();

        List<MParticleUser> users = new ArrayList<MParticleUser>();
        for (int i = 0; i < 10; i++) {
            users.add(new MockUser(ran.nextLong(), ran.nextLong()));
        }

        identityApi.sortUsers(users);

        long lastTime = Long.MAX_VALUE;
        for (MParticleUser user: users) {
            assertTrue(user.getLastSeenTime() <= lastTime);
            lastTime = user.getLastSeenTime();
        }
    }

    @Test
    public void testAliasUsersValidationRejection() {
        IdentityApi identityApi = new IdentityApi();
        identityApi.mMessageManager = new MessageManager() {
            @Override
            public void logAliasRequest(AliasRequest aliasRequest) {
                fail("should not logged Alias Request:\n" + aliasRequest.toString());
            }
        };

        //missing previousMpid
        AliasRequest request = AliasRequest.builder()
                .destinationMpid(123)
                .startTime(1)
                .endTime(2)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        //missing newMpid
        request = AliasRequest.builder()
                .sourceMpid(123)
                .startTime(1)
                .endTime(2)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        //newMpid and previousMpid are not unique
        request = AliasRequest.builder()
                .destinationMpid(123)
                .sourceMpid(123)
                .startTime(1)
                .endTime(2)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        //endTime before startTime
        request = AliasRequest.builder()
                .destinationMpid(123)
                .sourceMpid(456)
                .startTime(2)
                .endTime(1)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        //endTime and or startTime do not exist
        request = AliasRequest.builder()
                .destinationMpid(1)
                .sourceMpid(2)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        request = AliasRequest.builder()
                .destinationMpid(1)
                .sourceMpid(2)
                .startTime(3)
                .build();
        assertFalse(identityApi.aliasUsers(request));

        request = AliasRequest.builder()
                .destinationMpid(1)
                .sourceMpid(2)
                .endTime(3).build();
        assertFalse(identityApi.aliasUsers(request));
    }

    @Test
    public void testAliasUsersValidationAcceptance() {
        IdentityApi identityApi = new IdentityApi();
        final AndroidUtils.Mutable<Boolean> requestMade = new AndroidUtils.Mutable(false);
        identityApi.mMessageManager = new MessageManager() {
            @Override
            public void logAliasRequest(AliasRequest aliasRequest) {
                requestMade.value = true;
            }
        };
        AliasRequest request = AliasRequest.builder()
                .destinationMpid(1)
                .sourceMpid(2)
                .startTime(3)
                .endTime(4)
                .build();
        assertTrue(identityApi.aliasUsers(request));
        assertTrue(requestMade.value);
    }


    class MockUser extends AbstractMParticleUser {
        long mpid, lastSeenTime;

        MockUser(long mpid, long lastSeenTime) {
            this.mpid = mpid;
            this.lastSeenTime = lastSeenTime;
        }

        @Override
        public long getLastSeenTime() {
            return lastSeenTime;
        }

        @Override
        public long getId() {
            return mpid;
        }
    }
}

File: android-core/src/test/kotlin/com/mparticle/identity/IdentityApiTest.kt
package com.mparticle.identity

import com.mparticle.internal.MessageManager
import com.mparticle.mock.AbstractMParticleUser
import com.mparticle.testutils.AndroidUtils
import org.junit.Assert
import org.junit.Test
import java.util.Random

class IdentityApiTest {
    private var ran = Random()

    // IdentityApi should order their users from greatest lastSeenTime to least
    @Test
    fun testUserOrder() {
        val identityApi = IdentityApi()
        val users: MutableList<MParticleUser> = ArrayList()
        for (i in 0..9) {
            users.add(MockUser(ran.nextLong(), ran.nextLong()))
        }
        identityApi.sortUsers(users)
        var lastTime = Long.MAX_VALUE
        for (user in users) {
            Assert.assertTrue(user.lastSeenTime <= lastTime)
            lastTime = user.lastSeenTime
        }
    }

    @Test
    fun testAliasUsersValidationRejection() {
        val identityApi = IdentityApi()
        identityApi.mMessageManager = object : MessageManager() {
            override fun logAliasRequest(aliasRequest: AliasRequest) {
                Assert.fail("should not logged Alias Request:\n$aliasRequest")
            }
        }

        // missing previousMpid
        var request = AliasRequest.builder()
            .destinationMpid(123)
            .startTime(1)
            .endTime(2)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))

        // missing newMpid
        request = AliasRequest.builder()
            .sourceMpid(123)
            .startTime(1)
            .endTime(2)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))

        // newMpid and previousMpid are not unique
        request = AliasRequest.builder()
            .destinationMpid(123)
            .sourceMpid(123)
            .startTime(1)
            .endTime(2)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))

        // endTime before startTime
        request = AliasRequest.builder()
            .destinationMpid(123)
            .sourceMpid(456)
            .startTime(2)
            .endTime(1)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))

        // endTime and or startTime do not exist
        request = AliasRequest.builder()
            .destinationMpid(1)
            .sourceMpid(2)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))
        request = AliasRequest.builder()
            .destinationMpid(1)
            .sourceMpid(2)
            .startTime(3)
            .build()
        Assert.assertFalse(identityApi.aliasUsers(request))
        request = AliasRequest.builder()
            .destinationMpid(1)
            .sourceMpid(2)
            .endTime(3).build()
        Assert.assertFalse(identityApi.aliasUsers(request))
    }

    @Test
    fun testAliasUsersValidationAcceptance() {
        val identityApi = IdentityApi()
        val requestMade = AndroidUtils.Mutable(false)
        identityApi.mMessageManager = object : MessageManager() {
            override fun logAliasRequest(aliasRequest: AliasRequest) {
                requestMade.value = true
            }
        }
        val request = AliasRequest.builder()
            .destinationMpid(1)
            .sourceMpid(2)
            .startTime(3)
            .endTime(4)
            .build()
        Assert.assertTrue(identityApi.aliasUsers(request))
        Assert.assertTrue(requestMade.value)
    }

    internal inner class MockUser(private var mpid: Long, private var mpLastSeenTime: Long) :
        AbstractMParticleUser() {
        override fun getLastSeenTime(): Long {
            return mpLastSeenTime
        }

        override fun getId(): Long {
            return mpid
        }
    }
}
